### ISSUE TITLE
[JOSS review] Tests

### ISSUE THREAD

**Issue by arokem** [bmcfee/resampy#38]
A couple of sub-issues here:

-  Why are tests in the top-level directory, rather than under the package? If tests are installed, users can easily verify that the software is properly installed, for example by running `nosetests resampy`. As it stands, users will need to download the source, build the extensions and then run the tests from the source tree.

- Having said that, even after building the package, I get the following test errors (maybe I'm doing it wrong?): 

```
D-173-250-236-94:resampy (master) $nosetests
EEEEEEEE.E.EEE
======================================================================
ERROR: test_core.test_shape
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/arokem/anaconda/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_shape() missing 1 required positional argument: 'axis'

======================================================================
ERROR: test_core.test_bad_sr
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/arokem/anaconda/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_bad_sr() missing 2 required positional arguments: 'sr_orig' and 'sr_new'

======================================================================
ERROR: test_core.test_bad_rolloff
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/arokem/anaconda/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_bad_rolloff() missing 1 required positional argument: 'rolloff'

======================================================================
ERROR: test_core.test_bad_precision
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/arokem/anaconda/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/Users/arokem/source/resampy/tests/test_core.py", line 41, in test_bad_precision
    resampy.resample(x, 100, 50, filter='sinc_window', precision=-1)
  File "/Users/arokem/source/resampy/resampy/core.py", line 98, in resample
    interp_win, precision, _ = get_filter(filter, **kwargs)
  File "/Users/arokem/source/resampy/resampy/filters.py", line 159, in get_filter
    return getattr(sys.modules[__name__], name_or_function)(**kwargs)
  File "/Users/arokem/source/resampy/resampy/filters.py", line 108, in sinc_window
    raise ValueError('Invalid precision: precision={}'.format(precision))
ValueError: Invalid precision: precision=-1

======================================================================
ERROR: test_core.test_bad_num_zeros
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/arokem/anaconda/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/Users/arokem/source/resampy/tests/test_core.py", line 47, in test_bad_num_zeros
    resampy.resample(x, 100, 50, filter='sinc_window', num_zeros=0)
  File "/Users/arokem/source/resampy/resampy/core.py", line 98, in resample
    interp_win, precision, _ = get_filter(filter, **kwargs)
  File "/Users/arokem/source/resampy/resampy/filters.py", line 159, in get_filter
    return getattr(sys.modules[__name__], name_or_function)(**kwargs)
  File "/Users/arokem/source/resampy/resampy/filters.py", line 105, in sinc_window
    raise ValueError('Invalid num_zeros: num_zeros={}'.format(num_zeros))
ValueError: Invalid num_zeros: num_zeros=0

======================================================================
ERROR: test_core.test_dtype
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/arokem/anaconda/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_dtype() missing 1 required positional argument: 'dtype'

======================================================================
ERROR: test_core.test_bad_window
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/arokem/anaconda/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/Users/arokem/source/resampy/tests/test_core.py", line 62, in test_bad_window
    resampy.resample(x, 100, 200, filter='sinc_window', window=np.ones(50))
  File "/Users/arokem/source/resampy/resampy/core.py", line 98, in resample
    interp_win, precision, _ = get_filter(filter, **kwargs)
  File "/Users/arokem/source/resampy/resampy/filters.py", line 159, in get_filter
    return getattr(sys.modules[__name__], name_or_function)(**kwargs)
  File "/Users/arokem/source/resampy/resampy/filters.py", line 99, in sinc_window
    raise TypeError('window must be callable, not type(window)={}'.format(type(window)))
TypeError: window must be callable, not type(window)=<class 'numpy.ndarray'>

======================================================================
ERROR: test_core.test_short_signal
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/arokem/anaconda/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/Users/arokem/source/resampy/tests/test_core.py", line 69, in test_short_signal
    resampy.resample(x, 4, 1)
  File "/Users/arokem/source/resampy/resampy/core.py", line 94, in resample
    'resample from {}->{}'.format(x.shape[axis], sr_orig, sr_new))
ValueError: Input signal length=2 is too small to resample from 4->1

======================================================================
ERROR: test_filters.test_filter_sinc
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/arokem/anaconda/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_filter_sinc() missing 5 required positional arguments: 'filt', 'window', 'num_zeros', 'precision', and 'rolloff'

======================================================================
ERROR: test_filters.test_filter_missing
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/arokem/source/resampy/resampy/filters.py", line 164, in get_filter
    return load_filter(name_or_function)
  File "/Users/arokem/source/resampy/resampy/filters.py", line 193, in load_filter
    data = np.load(pkg_resources.resource_filename(__name__, fname))
  File "/Users/arokem/anaconda/lib/python3.5/site-packages/numpy/lib/npyio.py", line 362, in load
    fid = open(file, "rb")
FileNotFoundError: [Errno 2] No such file or directory: '/Users/arokem/source/resampy/resampy/data/bad name.npz'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/arokem/anaconda/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/Users/arokem/source/resampy/tests/test_filters.py", line 41, in test_filter_missing
    resampy.filters.get_filter('bad name')
  File "/Users/arokem/source/resampy/resampy/filters.py", line 167, in get_filter
    '{}'.format(name_or_function))
NotImplementedError: Cannot load filter definition for bad name

======================================================================
ERROR: test_quality.test_quality_sine
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/arokem/anaconda/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_quality_sine() missing 4 required positional arguments: 'sr_orig', 'sr_new', 'fil', and 'rms'

======================================================================
ERROR: test_quality.test_quality_sweep
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/arokem/anaconda/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
TypeError: test_quality_sweep() missing 4 required positional arguments: 'sr_orig', 'sr_new', 'fil', and 'rms'

----------------------------------------------------------------------
Ran 14 tests in 0.348s

FAILED (errors=12)
```

**Comment by bmcfee** [comment]
Thanks for the review/comments @arokem!  Some responses:

> Why are tests in the top-level directory, rather than under the package?

I suppose this is a habit I picked up from developing packages with large regression test dependencies.  More generally, I prefer to keep tests separate from the package so that the tests operate in an environment that best matches application code (ie, package code is loaded by absolute import instead of relative).

This issue came up previously in building deb packages #36.  I'm open to moving the tests in this package, but haven't yet done so.



> Having said that, even after building the package, I get the following test errors (maybe I'm doing it wrong?):

The tests are written for `py.test`, not `nosetests`, so that probably explains the errors you're getting.

I guess this means the documentation should be expanded to include running tests. :grin:

**Comment by arokem** [comment]
Gotcha. 

Yes: documentation about testing would be good. 

But your call about moving the tests. Feel free to close this issue.

**Comment by bmcfee** [comment]
Okay.  I think I'll just update the documentation for now, and revisit test migration at the next release.  Thanks!

**Comment by bmcfee** [comment]
Fixed by merging #42
